docs: improve local development documentation (#7631)
renovatebot/renovate: docs/development/local-development.md

# Local Development

This document gives tips and tricks on how to run Renovate locally to add features or fix bugs.
You can improve this documentation by opening a pull request.
For example, if you think anything is unclear, or you think something needs to be added, open a pull request!

## Installation

> #### Prerequisites

Reviewer: ```suggestion
### Prerequisites
```

should be h3 ? 🤔

Author: The original docs jump from `h2` to `h4` as well, so I kept it that way. But we can change this now of-course... :smile:

> - Python `^3.7`

Author: I noticed we use Python 3.8 in the build GitHub Action:
https://github.com/renovatebot/renovate/blob/6334c03724cb0b17ebd47fa76c24752bfd030fba/.github/workflows/build.yml#L30

Should we update the list of requirements?

```suggestion
- Python `^3.8`
```

Reviewer: sure

#### Fork and Clone

If you want to contribute to the project, you should first "fork" the main project using the GitHub Website and then clone your fork locally.
The Renovate project uses the [Yarn](https://github.com/yarnpkg/yarn) package management system instead of npm.

To ensure everything is working properly on your end, you must:

1. Install all dependencies with `yarn install`.
1. Make a build with `yarn build`, which should pass with no errors.
1. Verify all tests pass and have 100% test coverage, by running `yarn test`.

> 1. Verify the installation by running `yarn start`. You must see this error: `Fatal error: No authentication found for platform https://api.github.com/ (github)`

Reviewer: This numbering works 👀

Author: > This numbering works :eyes:

I'm not sure what you mean here? Do you mean I'm not breaking anything?

Or are you surprised that this kind of numbering scheme works in Markdown?

```markdown
1. ordered list item 1.
1. ordered list item 2.
1. ordered list item 3.
```

Markdown just wants you to use a number for a ordered list, if you use `1.`'s it will generate the right numbering sequence.
This way you can re-arrange the ordered list without having to change the numbering scheme.

You can use 1. 2. 3. though, that also works, but it's more work when you want to edit the order later.

Reviewer: Just was a little suprised. New stuff learned 😅

You only need to do these 4 steps this one time.

Before you submit a pull request you should:

1. Install newer dependencies with `yarn install`.
1. Run the tests with `yarn test`.

## Platform Account Setup

Although it's possible to make small source code improvements without testing against a real repository, in most cases you should run a "real" test on a repository before you submit a feature or fix.
It's possible to do this against GitHub, GitLab or Bitbucket public servers.

> #### Register new account (optional)

Author: ```suggestion
### Register new account (optional)
```
Should we change this as well then?

## Tests

You can run `yarn test` locally to test your code.
We test all PRs using the same tests, run on GitHub Actions.
`yarn test` runs an `eslint` check, a `prettier` check, a `type` check and then all the unit tests using `jest`.

### Prerequisites

You need to have Python with `mock` installed for all tests to pass.

> Python 3 includes `mock` so that approach is recommended.

Reviewer: I think this can be removed, as python 3.7 has that buildin and we force >=3.7

Maybe the next sentense can be mored somewhere else and we remove this section completely

Author: I've moved the sentence about the `.npmrc` file up into the fork and clone steps so that it's more visible.

You also need to make sure that you don't have a local `.npmrc` file that overrides npm's default registry.

### Jest

You can run just the Jest unit tests by running `yarn jest`.
You can also run just a subset of the Jest tests using file matching, e.g. `yarn jest composer` or `yarn jest workers/branch`.
If you get a test failure due to a "snapshot" mismatch, and you are sure that you need to update the snapshot, then you can append `-u` to the end.

> e.g. `yarn jest composer -u` would update the saved snapshots for _all_ tests in `test/manager/composer/*`.

Reviewer: ```suggestion
e.g. `yarn jest composer -u` would update the saved snapshots for _all_ tests in `**/composer/**`.
```

## Tests

You can run `yarn test` locally to test your code.

> We test all PRs using the same tests, run on CircleCI and Azure Pipelines.

Reviewer: ```suggestion
We test all PRs using the same tests, run on GitHub Actions.
```
